### PR TITLE
update openssl to v1.1.x

### PR DESCRIPTION
**Pull request by zhaozg**

### CODE REVIEW

**Comment by zhaozg** [comment]
ref https://github.com/luvit/lit/issues/256#issuecomment-478288221

**Comment by zhaozg** [comment]
https://github.com/luvit/luvit/issues/1060

File: deps/tls/common.lua [hunk 1]
```diff
@@ -26,9 +26,60 @@ local uv = require('uv')
...
+
+  local _, _, V = openssl.version(true)
+  return V > 0x10100000
+end
+
+DEFAULT_CIPHERS = 'TLS_AES_256_GCM_SHA384:TLS_AES_128_GCM_SHA256:' .. --TLS 1.3
+                  'ECDHE-RSA-AES128-SHA256:AES128-GCM-SHA256:' ..     --TLS 1.2
+                  'RC4:HIGH:!MD5:!aNULL:!EDH'                         --TLS 1.0
+
+if isTLSv1_3 then
```
Comment (squeek502) [inline deps/tls/common.lua]: Shouldn't this be `isTLSv1_3()`?

**Comment by zhaozg** [comment]
Not finished, need to do more for openssl v1.1.1 and TLSv1_3

File: deps/tls/common.lua [hunk 2]
```diff
@@ -26,9 +26,59 @@ local uv = require('uv')
...
+local isTLSv1_3 = function()
+  if isLibreSSL() then
+    return false
+  end
+
+  local _, _, V = openssl.version(true)
+  return V > 0x10100000
+end
+
+DEFAULT_CIPHERS = 'ECDHE-RSA-AES128-SHA256:AES128-GCM-SHA256:' ..     --TLS 1.2
```
Comment (squeek502) [inline deps/tls/common.lua]: Did you intentionally remove the TLS 1.3 cipher when you force pushed the new commit?

Comment (zhaozg) [reply deps/tls/common.lua]: When I try to use tls1_3 secureprotocol,  report ssl error 'no share cipher', after test ok, I'll set right cipher back

Comment (zhaozg) [reply deps/tls/common.lua]: When I try to use tls1_3 secureprotocol,  report ssl error 'no share cipher', after test ok, I'll set right cipher back

**Comment by zhaozg** [comment]
related: https://github.com/openssl/openssl/issues/8624

**Comment by zhaozg** [comment]
Done. ready to merge.

**Comment by zhaozg** [comment]
related: https://github.com/openssl/openssl/pull/6944

**Comment by zhaozg** [comment]
A few time, test will got 
```
Uncaught Error: /Users/zhaozg/work/luvit/luvit/tests/test-tls-remote.lua:34: EINVAL: invalid argument
stack traceback:
	[C]: in function 'assert'
	/Users/zhaozg/work/luvit/luvit/tests/test-tls-remote.lua:34: in function 'connectionListener'
	/Users/zhaozg/work/luvit/luvit/deps/tls/init.lua:53: in function 'handler'
	/Users/zhaozg/work/luvit/luvit/deps/core.lua:248: in function 'emit'
	/Users/zhaozg/work/luvit/luvit/deps/tls/common.lua:261: in function 'handler'
	/Users/zhaozg/work/luvit/luvit/deps/core.lua:248: in function 'emit'
	/Users/zhaozg/work/luvit/luvit/deps/tls/common.lua:432: in function 'handshake'
	/Users/zhaozg/work/luvit/luvit/deps/tls/common.lua:403: in function </Users/zhaozg/work/luvit/luvit/deps/tls/common.lua:392>
	[C]: in function 'run'
	/Users/zhaozg/work/luvit/luvit/tests/libs/tap.lua:79: in function </Users/zhaozg/work/luvit/luvit/tests/libs/tap.lua:64>
```
That because Socket:address() return error EINVAL: invalid argument
I double check
https://github.com/luvit/luv/blob/master/src/tcp.c#L144-L152, and can't finger the fail reason.
I think that can be solve in others PR.
This PR is DONE, ready to merge.

**Comment by zhaozg** [comment]
image

**Comment by squeek502** [comment]
Getting the following error in `test-tls-remote.lua`:

```
# Starting Test: tls remote-address
{ message = 'unable to get local issuer certificate' }
Uncaught Error: /home/ryan/Documents/luvit/tests/test-tls-remote.lua:34: ENOTCONN: socket is not connected
stack traceback:
	[C]: in function 'assert'
	/home/ryan/Documents/luvit/tests/test-tls-remote.lua:34: in function 'connectionListener'
	/home/ryan/Documents/luvit/deps/tls/init.lua:53: in function 'handler'
	/home/ryan/Documents/luvit/deps/core.lua:248: in function 'emit'
	/home/ryan/Documents/luvit/deps/tls/common.lua:261: in function 'handler'
	/home/ryan/Documents/luvit/deps/core.lua:248: in function 'emit'
	/home/ryan/Documents/luvit/deps/tls/common.lua:432: in function 'handshake'
	/home/ryan/Documents/luvit/deps/tls/common.lua:403: in function </home/ryan/Documents/luvit/deps/tls/common.lua:392>
	[C]: in function 'run'
	/home/ryan/Documents/luvit/tests/libs/tap.lua:79: in function </home/ryan/Documents/luvit/tests/libs/tap.lua:64>
```

The `{ message = 'unable to get local issuer certificate' }` is coming from `c:on('error', function(err)` (line 50-53 of `test-tls-remote.lua`)

I'm on 32-bit Ubuntu 16.04 
```
luvit version: 2.15.0
luvi version: v2.9.3
rex version: 8.37 2015-04-28
libuv version: 1.27.0
ssl version: OpenSSL 1.1.1b  26 Feb 2019, lua-openssl 0.7.5
```

**Comment by rphillips** [comment]
I just tried this successfully on OSX.

**Comment by zhaozg** [comment]
@philips After this merge, I meet a git hole to make other  PR. @

**Comment by zhaozg** [comment]
@squeek502 please check test-tls-remote.lua with https://github.com/luvit/luvit/pull/1065/commits/6159e11d5d0bb356c954e88908c93cc1a39966e9

**Comment by squeek502** [comment]
@zhaozg that fixes the test but it's still concerning that the test was passing on Mac/Windows but failed on Linux. Is that a regression in Libuv? If so, we should report it there.

**Comment by zhaozg** [comment]
fail because s:address() called to later(in ssl) after tcp close, move it to server accpet tcp and works fine. not related with os or platfirm

**Comment by squeek502** [comment]
But why did it work on Mac and Windows?

**Comment by zhaozg** [comment]
I think because TLS v1.3

**Comment by zhaozg** [comment]
and linux kernel handle syscall than other,so   linux close fd, but windows not finish close

**Pull request merged**